tests/integration/data_sources_and_expectations/expectations/test_expect_query_results_to_match_source.py

```diff
@@ -0,0 +1,326 @@
+        )
+    )
+    assert result.success
+
+
+@multi_source_batch_setup(
+    multi_source_test_configs=ALL_SOURCE_TO_TARGET_SOURCES,
+    target_data=TARGET_DATA_WITH_DUPS,
+    source_data=SOURCE_DATA_WITH_DUPS,
+)
+def test_expect_query_results_to_match_source_dups_failure(multi_source_batch: MultiSourceBatch):
+    result = multi_source_batch.target_batch.validate(
+        gxe.ExpectQueryResultsToMatchSource(
+            target_query="SELECT * FROM {batch} ORDER BY a DESC LIMIT 3",  # exclude one dup
+            source_data_source_name=multi_source_batch.source_data_source_name,
+            source_query=f"SELECT * FROM {multi_source_batch.source_table_name}",
+        )
+    )
+    assert not result.success
+    assert not result.exception_info["raised_exception"]
+
+
+@pytest.mark.parametrize(
+    "target_query,source_query,unexpected_percent,unexpected_count",
+    [
+        pytest.param(
+            "SELECT * FROM {batch} ORDER BY a LIMIT 100",  # 100 records
+            "SELECT * FROM {source_table} WHERE a >= 100 ORDER BY a",  # 100 records
+            0,
+            0,
+            id="only_match",
+        ),
+        pytest.param(
+            "SELECT * FROM {batch}",  # 200 records (half match)
+            "SELECT * FROM {source_table} WHERE a >= 100",  # 100 records
+            50,
+            100,
+            id="match_and_unexpected",
+        ),
+        pytest.param(
+            "SELECT * FROM {batch} ORDER BY a LIMIT 100",  # 100 records
+            "SELECT * FROM {source_table}",  # 200 records (half match)
+            50,
+            100,
+            id="match_and_missing",
+        ),
+        pytest.param(
+            "SELECT * FROM {batch} ORDER BY a DESC LIMIT 100",  # 100 different records
+            "SELECT * FROM {source_table} ORDER BY a LIMIT 100",  # 100 records
+            100,
+            100,
+            id="missing_and_unexpected",
+        ),
+        pytest.param(
+            "SELECT * FROM {batch}",  # 200 records (half match)
+            "SELECT * FROM {source_table}",  # 200 records
+            50,
+            100,
+            id="match_and_missing_and_unexpected",
+        ),
+        pytest.param(
+            "SELECT * FROM {batch}",  # 200 records
+            "SELECT * FROM {source_table} ORDER BY a LIMIT 100",  # 100 different records
+            100,
+            200,
+            id="only_unexpected",
+        ),
+        pytest.param(
+            "SELECT * FROM {batch} ORDER BY a DESC LIMIT 100",  # 100 records
+            "SELECT * FROM {source_table} LIMIT 0",  # 0 records
+            100,
+            100,
+            id="only_missing",
+        ),
+        pytest.param(
+            "SELECT * FROM {batch} LIMIT 0",  # 0 records
+            "SELECT * FROM {source_table} LIMIT 0",  # 0 records
+            0,
+            0,
+            id="nothing_to_compare",
+        ),
+    ],
+)
+@multi_source_batch_setup(
```

Comment:
These test cases all look good, but I think there's a bug around how we handle duplicates currently (see the comments in the implementation). I think we need a test or two around that.

Reply:
I have a dups test, so I will just add assertions for `unexpected_percent` there.

Reply:
FWIW I do think there's a lot of value in having one single parameterized test that covers all the different math stuff. But that's non-blocking for me